xds/src/main/java/io/grpc/xds/XdsClientImpl2.java

```diff
@@ -0,0 +1,1688 @@
+      rdsResourceSubscribers.remove(resourceName);
+      adjustResourceSubscription(ResourceType.RDS, rdsResourceSubscribers.keySet());
+    }
+  }
+
+  @Override
+  void watchCdsResource(String resourceName, CdsResourceWatcher watcher) {
+    ResourceSubscriber subscriber = cdsResourceSubscribers.get(resourceName);
+    if (subscriber == null) {
+      logger.log(XdsLogLevel.INFO, "Subscribe CDS resource {0}", resourceName);
+      subscriber = new ResourceSubscriber(ResourceType.CDS, resourceName);
+      cdsResourceSubscribers.put(resourceName, subscriber);
+      adjustResourceSubscription(ResourceType.CDS, cdsResourceSubscribers.keySet());
+    }
+    subscriber.addWatcher(watcher);
+  }
+
+  @Override
+  void cancelCdsResourceWatch(String resourceName, CdsResourceWatcher watcher) {
+    ResourceSubscriber subscriber = cdsResourceSubscribers.get(resourceName);
+    subscriber.removeWatcher(watcher);
+    if (!subscriber.isWatched()) {
+      subscriber.stopTimer();
+      logger.log(XdsLogLevel.INFO, "Unsubscribe CDS resource {0}", resourceName);
+      cdsResourceSubscribers.remove(resourceName);
+      adjustResourceSubscription(ResourceType.CDS, cdsResourceSubscribers.keySet());
+    }
+  }
+
+  @Override
+  void watchEdsResource(String resourceName, EdsResourceWatcher watcher) {
+    ResourceSubscriber subscriber = edsResourceSubscribers.get(resourceName);
+    if (subscriber == null) {
+      logger.log(XdsLogLevel.INFO, "Subscribe EDS resource {0}", resourceName);
+      subscriber = new ResourceSubscriber(ResourceType.EDS, resourceName);
+      edsResourceSubscribers.put(resourceName, subscriber);
+      adjustResourceSubscription(ResourceType.EDS, edsResourceSubscribers.keySet());
+    }
+    subscriber.addWatcher(watcher);
+  }
+
+  @Override
+  void cancelEdsResourceWatch(String resourceName, EdsResourceWatcher watcher) {
+    ResourceSubscriber subscriber = edsResourceSubscribers.get(resourceName);
+    subscriber.removeWatcher(watcher);
+    if (!subscriber.isWatched()) {
+      subscriber.stopTimer();
+      logger.log(XdsLogLevel.INFO, "Unsubscribe EDS resource {0}", resourceName);
+      edsResourceSubscribers.remove(resourceName);
+      adjustResourceSubscription(ResourceType.EDS, edsResourceSubscribers.keySet());
+    }
+  }
+
+  @Override
+  void watchListenerData(int port, ListenerWatcher watcher) {
+    checkState(listenerWatcher == null, "ListenerWatcher already registered");
+    listenerWatcher = checkNotNull(watcher, "watcher");
+    checkArgument(port > 0, "port needs to be > 0");
+    this.listenerPort = port;
+    logger.log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
+    if (rpcRetryTimer != null && rpcRetryTimer.isPending()) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    updateNodeMetadataForListenerRequest(port);
+    adsStream.sendXdsRequest(ResourceType.LDS, ImmutableList.<String>of());
```

Comment:
Seems this request is never retried for the `listenerWatcher` case.

Reply:
Forgot that was for Sanjay's code. Added back.